### PR TITLE
Fix object file type autodetection (issue #2105)

### PR DESCRIPTION
**Pull request by kinke**

### CODE REVIEW

**Comment by kinke** [comment]
@JohanEngelen: Are you fine with this or do you insist on a test? ;)

File: driver/main.cpp [hunk 1]
```diff
@@ -530,29 +530,24 @@ void parseCommandLine(int argc, char **argv, Strings &sourceFiles,
 
   // LDC output determination
 
-  // if we don't link, autodetect target from extension
-  if (!global.params.link && !global.params.lib && global.params.objname) {
+  // if we don't link and there's no `-output-*` switch but an `-of` one,
+  // autodetect type of desired 'object' file from file extension
+  if (!global.params.link && !global.params.lib && global.params.objname &&
+      global.params.output_o == OUTPUTFLAGdefault) {
```
Comment (kinke) [inline driver/main.cpp]: Some more context: `global.params.output_o` equals `OUTPUTFLAGdefault` here if `-output-o` is omitted (neither enabled nor disabled) AND none of the others are enabled (so omitted or explicitly disabled).

Comment (JohanEngelen) [inline driver/main.cpp]: Thanks, this helped :)

**Comment by JohanEngelen** [comment]
So if I understand correctly, with this PR:
`ldc2 test.d -output-ll -of=test.s` will only output IR to the file `test.s`, right?

Yeah, tests would help. Tedious, I know, so perhaps you can spend the time elsewhere.

**Comment by kinke** [comment]
> ldc2 test.d -output-ll -of=test.s will only output IR to the file test.s, right?

Yep, but that's no change in behavior, that was already the case, just verified with 1.3.0-beta1.

**Comment by JohanEngelen** [comment]
Ah I meant with `-c`. (surprised things works without `-c` !).
So: `ldc2 test.d -c -output-ll -of=test.s`.  If that only outputs `test.s` with LLVM IR, I'm happy.

**Comment by kinke** [comment]
> (surprised things works without -c !).

I used it a couple of times in lit tests IIRC. It makes sense though, doesn't it? Only write IR to the output file, skip codegen & linking.
Yep, adding `-c` makes no difference with this PR anymore if there's at least one (non-disabled) `-output-*` option, just verified. ;)

**Comment by JohanEngelen** [comment]
Clang is complaining about the code:
```
../driver/main.cpp:546:32: warning: comparison of constant 'BOU_FALSE' (2) with expression of type 'bool' is always true [-Wtautological-constant-out-of-range-compare]
    } else if (opts::output_ll != cl::BOU_FALSE &&
               ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~
../driver/main.cpp:550:32: warning: comparison of constant 'BOU_FALSE' (2) with expression of type 'bool' is always true [-Wtautological-constant-out-of-range-compare]
    } else if (opts::output_bc != cl::BOU_FALSE &&
               ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~
../driver/main.cpp:554:31: warning: comparison of constant 'BOU_FALSE' (2) with expression of type 'bool' is always true [-Wtautological-constant-out-of-range-compare]
    } else if (opts::output_s != cl::BOU_FALSE &&
               ~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~
```

**Comment by kinke** [comment]
Thanks Johan, fixed in 6c6379d534ef44cf38cafed5707f0d39310e820d.

**Pull request merged**